File: ios/Shared/NativeBrokerTransport.swift
import Foundation

/// The closed error-kind set exported by the broker binding, plus the two bounded platform
/// failures that can occur before a usable native result exists.
public enum BrokerNativeFailureKind: String, CaseIterable, Sendable {
    case cancelled
    case timeout
    case rateLimited = "rate_limited"
    case httpStatus = "http_status"
    case dns
    case tls
    case network
    case verification
    case validation
    case unknown
    case unavailable
    case decode

    init(bindingValue: String) {
        switch bindingValue.trimmingCharacters(in: .whitespacesAndNewlines) {
        case BrokerNativeFailureKind.cancelled.rawValue:
            self = .cancelled
        case BrokerNativeFailureKind.timeout.rawValue:
            self = .timeout
        case BrokerNativeFailureKind.rateLimited.rawValue:
            self = .rateLimited
        case BrokerNativeFailureKind.httpStatus.rawValue:
            self = .httpStatus
        case BrokerNativeFailureKind.dns.rawValue:
            self = .dns
        case BrokerNativeFailureKind.tls.rawValue:
            self = .tls
        case BrokerNativeFailureKind.network.rawValue:
            self = .network
        case BrokerNativeFailureKind.verification.rawValue:
            self = .verification
        case BrokerNativeFailureKind.validation.rawValue:
            self = .validation
        case BrokerNativeFailureKind.unknown.rawValue:
            self = .unknown
        default:
            // Platform-only kinds are never accepted from the binding. Blank and future values
            // stay in the bounded unknown bucket.
            self = .unknown
        }
    }
}

/// A bounded, platform-owned projection of a failed broker binding result.
///
/// `message` is diagnostic only. Callers make policy decisions from `kind`, `httpStatus`, and
/// `retryAfterMilliseconds`; they must never parse the message.
public struct BrokerNativeFailure: LocalizedError, Equatable, Sendable {
    public let kind: BrokerNativeFailureKind
    public let httpStatus: Int?
    public let retryAfterMilliseconds: UInt64?
    public let message: String

    public init(
        kind: BrokerNativeFailureKind,
        httpStatus: Int? = nil,
        retryAfterMilliseconds: UInt64? = nil,
        message: String = ""
    ) {
        self.kind = kind
        self.httpStatus = httpStatus.flatMap { $0 > 0 ? $0 : nil }
        self.retryAfterMilliseconds = retryAfterMilliseconds.flatMap { $0 > 0 ? $0 : nil }
        self.message = Self.sanitize(message)
    }

    public init(
        bindingKind: String,
        httpStatus: Int32,
        retryAfterMilliseconds: Int64,
        message: String
    ) {
        self.init(
            kind: BrokerNativeFailureKind(bindingValue: bindingKind),
            httpStatus: httpStatus > 0 ? Int(httpStatus) : nil,
            retryAfterMilliseconds: retryAfterMilliseconds > 0
                ? UInt64(retryAfterMilliseconds)
                : nil,
            message: message
        )
    }

    public var errorDescription: String? {
        message.isEmpty ? "The native broker request failed (\(kind.rawValue))." : message
    }

    /// Local schema/configuration/linkage failures cannot improve by minting another WSS ticket.
    public var isLocalPlatformFailure: Bool {
        switch kind {
        case .validation, .unavailable, .decode:
            return true
        default:
            return false
        }
    }

    /// Existing bounded mobile dashboard taxonomy. Binding strings never become telemetry values.
    public var failureReason: String {
        switch kind {
        case .cancelled:
            return "cancelled"
        case .timeout:
            return "timeout"
        case .rateLimited:
            return "rate_limited"
        case .httpStatus:
            if httpStatus == 429 { return "rate_limited" }
            return httpStatus.map { "http_\($0)" } ?? "unknown"
        case .dns:
            return "dns_failure"
        case .tls:
            return "tls_handshake"
        case .network:
            return "network_unreachable"
        case .verification, .validation, .unknown, .unavailable, .decode:
            return "unknown"
        }
    }

    static func sanitize(_ value: String, maxBytes: Int = 256) -> String {
        let scalars = value.unicodeScalars.map { scalar -> Character in
            CharacterSet.controlCharacters.contains(scalar) ? " " : Character(scalar)
        }
        let collapsed = String(scalars)
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
        let bytes = Array(collapsed.utf8)
        guard bytes.count > maxBytes else { return collapsed }
        var end = maxBytes
        while end > 0, (bytes[end] & 0xC0) == 0x80 {
            end -= 1
        }
        return String(decoding: bytes[..<end], as: UTF8.self)
    }
}

/// Common fields copied from every generated broker result before its operation is closed.
public protocol NativeBrokerResult: Sendable {
    var succeeded: Bool { get }
    var errorKind: String { get }
    var errorText: String { get }
    var httpStatus: Int32 { get }
    var retryAfterMilliseconds: Int64 { get }
}

public struct NativeBrokerResultSnapshot: NativeBrokerResult, Equatable, Sendable {
    public let succeeded: Bool
    public let errorKind: String
    public let errorText: String
    public let httpStatus: Int32
    public let retryAfterMilliseconds: Int64

    public init(
        succeeded: Bool,
        errorKind: String = "",
        errorText: String = "",
        httpStatus: Int32 = 0,
        retryAfterMilliseconds: Int64 = 0
    ) {
        self.succeeded = succeeded
        self.errorKind = errorKind
        self.errorText = BrokerNativeFailure.sanitize(errorText)
        self.httpStatus = httpStatus
        self.retryAfterMilliseconds = retryAfterMilliseconds
    }
}

public struct NativeBrokerRelayResultSnapshot:
    NativeBrokerResult,
    CustomStringConvertible,
    CustomDebugStringConvertible,
    Equatable,
    Sendable
{
    public let succeeded: Bool
    public let errorKind: String
    public let errorText: String
    public let httpStatus: Int32
    public let retryAfterMilliseconds: Int64
    public let brokerURL: String
    public let relayJSON: String
    public let keyID: String
    public let signatureVerified: Bool

    public init(
        succeeded: Bool,
        errorKind: String = "",
        errorText: String = "",
        httpStatus: Int32 = 0,
        retryAfterMilliseconds: Int64 = 0,
        brokerURL: String = "",
        relayJSON: String = "",
        keyID: String = "",
        signatureVerified: Bool = false
    ) {
        self.succeeded = succeeded
        self.errorKind = errorKind
        self.errorText = BrokerNativeFailure.sanitize(errorText)
        self.httpStatus = httpStatus
        self.retryAfterMilliseconds = retryAfterMilliseconds
        self.brokerURL = brokerURL
        self.relayJSON = relayJSON
        self.keyID = keyID
        self.signatureVerified = signatureVerified
    }

    public var description: String {
        "NativeBrokerRelayResultSnapshot(succeeded: \(succeeded), " +
            "errorKind: \(errorKind), httpStatus: \(httpStatus), " +
            "retryAfterMilliseconds: \(retryAfterMilliseconds), brokerURL: <redacted>, " +
            "relayJSON: <redacted>, keyID: \(keyID), " +
            "signatureVerified: \(signatureVerified))"
    }

    public var debugDescription: String { description }
}

public struct NativeBrokerSpeedTestResultSnapshot: NativeBrokerResult, Equatable, Sendable {
    public let succeeded: Bool
    public let errorKind: String
    public let errorText: String
    public let httpStatus: Int32
    public let retryAfterMilliseconds: Int64
    public let bytes: Int64
    public let timeToFirstByteMilliseconds: Int64
    public let downloadDurationMilliseconds: Int64
    public let totalDurationMilliseconds: Int64
    public let megabitsPerSecond: Double

    public init(
        succeeded: Bool,
        errorKind: String = "",
        errorText: String = "",
        httpStatus: Int32 = 0,
        retryAfterMilliseconds: Int64 = 0,
        bytes: Int64 = 0,
        timeToFirstByteMilliseconds: Int64 = 0,
        downloadDurationMilliseconds: Int64 = 0,
        totalDurationMilliseconds: Int64 = 0,
        megabitsPerSecond: Double = 0
    ) {
        self.succeeded = succeeded
        self.errorKind = errorKind
        self.errorText = BrokerNativeFailure.sanitize(errorText)
        self.httpStatus = httpStatus
        self.retryAfterMilliseconds = retryAfterMilliseconds
        self.bytes = bytes
        self.timeToFirstByteMilliseconds = timeToFirstByteMilliseconds
        self.downloadDurationMilliseconds = downloadDurationMilliseconds
        self.totalDurationMilliseconds = totalDurationMilliseconds
        self.megabitsPerSecond = megabitsPerSecond
    }
}

public struct NativeBrokerManifestResultSnapshot:
    NativeBrokerResult,
    Equatable,
    Sendable,
    CustomStringConvertible,
    CustomDebugStringConvertible
{
    public let succeeded: Bool
    public let errorKind: String
    public let errorText: String
    public let httpStatus: Int32
    public let retryAfterMilliseconds: Int64
    public let bodyJSON: String
    public let sourceURL: String

    public init(
        succeeded: Bool,
        errorKind: String = "",
        errorText: String = "",
        httpStatus: Int32 = 0,
        retryAfterMilliseconds: Int64 = 0,
        bodyJSON: String = "",
        sourceURL: String = ""
    ) {
        self.succeeded = succeeded
        self.errorKind = errorKind
        self.errorText = BrokerNativeFailure.sanitize(errorText)
        self.httpStatus = httpStatus
        self.retryAfterMilliseconds = retryAfterMilliseconds
        self.bodyJSON = bodyJSON
        self.sourceURL = sourceURL
    }

    public var description: String {
        "NativeBrokerManifestResultSnapshot(succeeded: \(succeeded), " +
            "errorKind: \(errorKind), httpStatus: \(httpStatus), " +
            "retryAfterMilliseconds: \(retryAfterMilliseconds), bodyJSON: <redacted>, " +
            "sourceURL: \(sourceURL))"
    }

    public var debugDescription: String { description }
}

public struct NativeBrokerWSSTicketResultSnapshot:
    NativeBrokerResult,
    Equatable,
    Sendable,
    CustomStringConvertible,
    CustomDebugStringConvertible
{
    public let succeeded: Bool
    public let errorKind: String
    public let errorText: String
    public let httpStatus: Int32
    public let retryAfterMilliseconds: Int64
    public let ticket: String
    public let url: String
    public let expiresAtMilliseconds: Int64

    public init(
        succeeded: Bool,
        errorKind: String = "",
        errorText: String = "",
        httpStatus: Int32 = 0,
        retryAfterMilliseconds: Int64 = 0,
        ticket: String = "",
        url: String = "",
        expiresAtMilliseconds: Int64 = 0
    ) {
        self.succeeded = succeeded
        self.errorKind = errorKind
        self.errorText = BrokerNativeFailure.sanitize(errorText)
        self.httpStatus = httpStatus
        self.retryAfterMilliseconds = retryAfterMilliseconds
        self.ticket = ticket
        self.url = url
        self.expiresAtMilliseconds = expiresAtMilliseconds
    }

    public var description: String {
        "NativeBrokerWSSTicketResultSnapshot(succeeded: \(succeeded), " +
            "errorKind: \(errorKind), httpStatus: \(httpStatus), " +
            "retryAfterMilliseconds: \(retryAfterMilliseconds), ticket: <redacted>, " +
            "url: <redacted>, expiresAtMilliseconds: \(expiresAtMilliseconds))"
    }

    public var debugDescription: String { description }
}

/// A single-use, synchronous native operation. Implementations copy generated results into the
/// snapshots above before returning. Hostless tests provide fakes and never load the Go runtime.
public protocol NativeBrokerOperation: AnyObject, Sendable {
    func firstReachable(
        primary: String,
        limit: Int32,
        clientID: String,
        sessionID: String
    ) -> NativeBrokerRelayResultSnapshot?

    func sendTelemetryBatchJSON(
        brokerURL: String,
        batchJSON: String
    ) -> NativeBrokerResultSnapshot?

    func runSpeedTest(
        brokerURL: String
    ) -> NativeBrokerSpeedTestResultSnapshot?

    func fetchManifestCandidate(
        candidateURL: String
    ) -> NativeBrokerManifestResultSnapshot?

    func requestWSSTicket(
        brokerURL: String,
        relayID: String,
        frontID: String,
        clientID: String,
        sessionID: String
    ) -> NativeBrokerWSSTicketResultSnapshot?

    /// May block until an in-flight Go request exits. It is idempotent and may run concurrently.
    func close()
}

public protocol NativeBrokerOperationFactory: Sendable {
    func makeOperation() -> (any NativeBrokerOperation)?
}

/// Constructor selection remains platform-owned and testable without importing Libbox.
///
/// Production implements this protocol in `LibboxBrokerTransport.swift`, after wrapping the
/// generated object in `NativeBrokerOperation`. Hostless tests use a spy constructor and therefore
/// never load the generated framework or a Go runtime.
protocol NativeBrokerOperationConstructing: Sendable {
    func makeIOSOperation(
        appVersion: String,
        osVersion: String
    ) -> (any NativeBrokerOperation)?

    func makeReactNativeOperation(
        appVersion: String,
        osToken: String
    ) -> (any NativeBrokerOperation)?
}

enum NativeBrokerOperationClient: Equatable, Sendable {
    case ios(osVersion: String)
    case reactNative(osToken: String)
}

struct ConfiguredNativeBrokerOperationFactory: NativeBrokerOperationFactory {
    let appVersion: String
    let client: NativeBrokerOperationClient
    let constructor: any NativeBrokerOperationConstructing

    /// Fixed platform selection for React Native callers. Keeping the token here makes the exact
    /// production constructor choice testable without importing or loading Libbox.
    static func forReactNativeIOS(
        appVersion: String,
        constructor: any NativeBrokerOperationConstructing
    ) -> Self {
        Self(
            appVersion: appVersion,
            client: .reactNative(osToken: "ios"),
            constructor: constructor
        )
    }

    func makeOperation() -> (any NativeBrokerOperation)? {
        switch client {
        case .ios(let osVersion):
            return constructor.makeIOSOperation(
                appVersion: appVersion,
                osVersion: osVersion
            )
        case .reactNative(let osToken):
            return constructor.makeReactNativeOperation(
                appVersion: appVersion,
                osToken: osToken
            )
        }
    }
}

/// Cancellation-safe bridge for gomobile's synchronous broker selectors.
///
/// The blocking selector runs on a detached worker. Cancellation schedules `close()` on a
/// different concurrent worker because Go `Close` waits for the selector to return. The selector
/// snapshots its result first; the detached worker then closes in all cases. A post-worker
/// cancellation check prevents a success racing with cancellation from committing caller state.
public enum NativeBrokerRunner {
    public static func run<Result: Sendable>(
        factory: any NativeBrokerOperationFactory,
        invoke: @escaping @Sendable (any NativeBrokerOperation) -> Result?
    ) async throws -> Result {
        try await runImpl(
            factory: factory,
            beforeWorkerStart: {},
            invoke: invoke
        )
    }

    /// Deterministic seam for exercising cancellation after the task pre-check but before the
    /// detached worker claims permission to invoke the synchronous selector.
    static func runForTesting<Result: Sendable>(
        factory: any NativeBrokerOperationFactory,
        beforeWorkerStart: @escaping @Sendable () async -> Void,
        invoke: @escaping @Sendable (any NativeBrokerOperation) -> Result?
    ) async throws -> Result {
        try await runImpl(
            factory: factory,
            beforeWorkerStart: beforeWorkerStart,
            invoke: invoke
        )
    }

    private static func runImpl<Result: Sendable>(
        factory: any NativeBrokerOperationFactory,
        beforeWorkerStart: @escaping @Sendable () async -> Void,
        invoke: @escaping @Sendable (any NativeBrokerOperation) -> Result?
    ) async throws -> Result {
        guard let operation = factory.makeOperation() else {
            throw BrokerNativeFailure(kind: .unavailable)
        }
        let startGate = NativeBrokerStartGate()

        return try await withTaskCancellationHandler {
            // A task can already be cancelled when it reaches the runner (for example while it
            // was queued behind another actor operation). Do not let that state mint a WSS ticket
            // or start any other native request. The cancellation handler still closes the
            // freshly-created operation asynchronously.
            try Task.checkCancellation()
            await beforeWorkerStart()
            let worker = Task.detached { () -> Result? in
                // Cancellation may land after the check above but before this worker executes.
                // Only one side wins the locked claim; a cancellation winner skips Go entirely.
                guard startGate.claimStart() else { return nil }
                defer { operation.close() }
                return invoke(operation)
            }
            let result = await worker.value
            try Task.checkCancellation()
            guard let result else {
                throw BrokerNativeFailure(kind: .unavailable)
            }
            return result
        } onCancel: {
            // Mark synchronously so a not-yet-started worker cannot race ahead of asynchronous
            // Close. Close itself must remain off this stack because Go waits for in-flight work.
            startGate.cancel()
            DispatchQueue.global(qos: .utility).async {
                operation.close()
            }
        }
    }
}

private final class NativeBrokerStartGate: @unchecked Sendable {
    private let lock = NSLock()
    private var cancelled = false
    private var startClaimed = false

    func cancel() {
        lock.lock()
        cancelled = true
        lock.unlock()
    }

    func claimStart() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard cancelled == false, startClaimed == false else { return false }
        startClaimed = true
        return true
    }
}

public extension NativeBrokerResult {
    func throwIfFailed() throws {
        guard succeeded == false else { return }
        let failure = BrokerNativeFailure(
            bindingKind: errorKind,
            httpStatus: httpStatus,
            retryAfterMilliseconds: retryAfterMilliseconds,
            message: errorText
        )
        if failure.kind == .cancelled {
            // Cancellation is only ever the caller's: a genuinely cancelled task keeps its
            // cooperative CancellationError here. A native "cancelled" kind reaching a live task
            // is a failed request — converting it into cancellation would silently end the
            // caller's epoch (skipping cleanup, failure status, and telemetry) on the strength
            // of a Go-side race or mis-mapped binding error.
            try Task.checkCancellation()
        }
        throw failure
    }
}

File: android/app/src/main/java/com/openrung/net/NativeBrokerTransport.kt
package com.openrung.net

import android.os.Build
import com.openrung.BuildConfig
import io.nekohasekai.libbox.Libbox
import io.nekohasekai.libbox.OpenRungBrokerManifestResult
import io.nekohasekai.libbox.OpenRungBrokerOperation
import io.nekohasekai.libbox.OpenRungBrokerRelayResult
import io.nekohasekai.libbox.OpenRungBrokerResult
import io.nekohasekai.libbox.OpenRungBrokerSpeedTestResult
import io.nekohasekai.libbox.OpenRungBrokerWSSTicketResult
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import java.io.IOException

/**
 * Bounded error kinds exposed to Android policy. The first ten values are the exact brokerapi
 * binding contract; [UNAVAILABLE] and [DECODE] describe failures introduced by this adapter.
 */
enum class BrokerNativeFailureKind(val value: String) {
    CANCELLED("cancelled"),
    TIMEOUT("timeout"),
    RATE_LIMITED("rate_limited"),
    HTTP_STATUS("http_status"),
    DNS("dns"),
    TLS("tls"),
    NETWORK("network"),
    VERIFICATION("verification"),
    VALIDATION("validation"),
    UNKNOWN("unknown"),
    UNAVAILABLE("unavailable"),
    DECODE("decode"),
    ;

    companion object {
        /** Blank and future binding values fail into the bounded [UNKNOWN] bucket. */
        fun fromBinding(value: String?): BrokerNativeFailureKind = when (value?.trim()) {
            CANCELLED.value -> CANCELLED
            TIMEOUT.value -> TIMEOUT
            RATE_LIMITED.value -> RATE_LIMITED
            HTTP_STATUS.value -> HTTP_STATUS
            DNS.value -> DNS
            TLS.value -> TLS
            NETWORK.value -> NETWORK
            VERIFICATION.value -> VERIFICATION
            VALIDATION.value -> VALIDATION
            UNKNOWN.value -> UNKNOWN
            else -> UNKNOWN
        }
    }
}

/**
 * A safe, bounded projection of a failed native request.
 *
 * [httpStatus] and [retryAfterMillis] are null when the generated getter returned zero. [message]
 * is built only from a fixed operation label and the binding's sanitized ErrorText; request JSON,
 * response bodies, result stringification, tickets, and returned WSS URLs are never attached.
 */
class BrokerNativeFailure(
    val kind: BrokerNativeFailureKind,
    val httpStatus: Int? = null,
    val retryAfterMillis: Long? = null,
    message: String,
    cause: Throwable? = null,
) : IOException(sanitizeBrokerFailureMessage(message), cause) {
    val isLocalPlatformFailure: Boolean
        get() = kind == BrokerNativeFailureKind.VALIDATION ||
            kind == BrokerNativeFailureKind.UNAVAILABLE ||
            kind == BrokerNativeFailureKind.DECODE
}

/** Common, platform-owned result shape copied out of gomobile before the operation is closed. */
interface NativeBrokerResult {
    val succeeded: Boolean
    val errorKind: String
    val errorText: String
    val httpStatus: Int
    val retryAfterMillis: Long
}

data class NativeBrokerCommonResult(
    override val succeeded: Boolean,
    override val errorKind: String = "",
    override val errorText: String = "",
    override val httpStatus: Int = 0,
    override val retryAfterMillis: Long = 0,
) : NativeBrokerResult

data class NativeBrokerRelayResult(
    override val succeeded: Boolean,
    override val errorKind: String = "",
    override val errorText: String = "",
    override val httpStatus: Int = 0,
    override val retryAfterMillis: Long = 0,
    val brokerUrl: String = "",
    val relayJson: String = "",
    val keyId: String = "",
    val signatureVerified: Boolean = false,
) : NativeBrokerResult {
    override fun toString(): String =
        "NativeBrokerRelayResult(" +
            "succeeded=$succeeded, errorKind=${BrokerNativeFailureKind.fromBinding(errorKind).value}, " +
            "httpStatus=$httpStatus, retryAfterMillis=$retryAfterMillis, " +
            "brokerUrl=$brokerUrl, relayJson=<redacted>, keyId=$keyId, " +
            "signatureVerified=$signatureVerified)"
}

data class NativeBrokerSpeedTestResult(
    override val succeeded: Boolean,
    override val errorKind: String = "",
    override val errorText: String = "",
    override val httpStatus: Int = 0,
    override val retryAfterMillis: Long = 0,
    val bytes: Long = 0,
    val ttfbMillis: Long = 0,
    val downloadDurationMillis: Long = 0,
    val totalDurationMillis: Long = 0,
    val mbps: Double = 0.0,
) : NativeBrokerResult

data class NativeBrokerManifestResult(
    override val succeeded: Boolean,
    override val errorKind: String = "",
    override val errorText: String = "",
    override val httpStatus: Int = 0,
    override val retryAfterMillis: Long = 0,
    val bodyJson: String = "",
    val sourceUrl: String = "",
) : NativeBrokerResult {
    override fun toString(): String =
        "NativeBrokerManifestResult(" +
            "succeeded=$succeeded, errorKind=${BrokerNativeFailureKind.fromBinding(errorKind).value}, " +
            "httpStatus=$httpStatus, retryAfterMillis=$retryAfterMillis, " +
            "bodyJson=<redacted>, sourceUrl=$sourceUrl)"
}

/**
 * Ticket snapshot with an explicitly redacted representation. The bearer and bound URL are still
 * available as fields for the immediate WSS handoff, but neither can leak through ordinary logs.
 */
class NativeBrokerWssTicketResult(
    override val succeeded: Boolean,
    override val errorKind: String = "",
    override val errorText: String = "",
    override val httpStatus: Int = 0,
    override val retryAfterMillis: Long = 0,
    val ticket: String = "",
    val url: String = "",
    val expiresAtMillis: Long = 0,
) : NativeBrokerResult {
    override fun toString(): String =
        "NativeBrokerWssTicketResult(" +
            "succeeded=$succeeded, errorKind=${BrokerNativeFailureKind.fromBinding(errorKind).value}, " +
            "httpStatus=$httpStatus, retryAfterMillis=$retryAfterMillis, " +
            "ticket=<redacted>, url=<redacted>, expiresAtMillis=$expiresAtMillis)"
}

/** One single-use operation. Tests implement this interface without loading gomobile or Go. */
interface NativeBrokerOperation {
    fun firstReachable(
        primary: String,
        limit: Int,
        clientId: String,
        sessionId: String,
    ): NativeBrokerRelayResult?

    fun sendTelemetryBatchJSON(
        brokerUrl: String,
        batchJson: String,
    ): NativeBrokerCommonResult?

    fun runSpeedTest(brokerUrl: String): NativeBrokerSpeedTestResult?

    fun fetchManifestCandidate(candidateUrl: String): NativeBrokerManifestResult?

    fun requestWSSTicket(
        brokerUrl: String,
        relayId: String,
        frontId: String,
        clientId: String,
        sessionId: String,
    ): NativeBrokerWssTicketResult?

    fun close()
}

fun interface NativeBrokerOperationFactory {
    fun create(): NativeBrokerOperation
}

/**
 * Platform-owned constructor seam: JVM tests can assert Android metadata without constructing a
 * generated object, while production alone delegates to Libbox's Android constructor.
 */
internal fun interface AndroidBrokerOperationConstructor {
    fun create(appVersion: String, apiLevel: String): NativeBrokerOperation?
}

internal class AndroidNativeBrokerOperationFactory(
    private val constructor: AndroidBrokerOperationConstructor = LibboxAndroidBrokerOperationConstructor,
    internal val appVersion: String = BuildConfig.VERSION_NAME,
    internal val apiLevel: String = Build.VERSION.SDK_INT.toString(),
) : NativeBrokerOperationFactory {
    override fun create(): NativeBrokerOperation {
        val operation = try {
            constructor.create(appVersion, apiLevel)
        } catch (error: LinkageError) {
            throw unavailableNativeBrokerFailure(error)
        }
        return operation ?: throw unavailableNativeBrokerFailure()
    }
}

/**
 * Separate constructor seam for React Native operations. The OS token is deliberately fixed by
 * the platform factory rather than accepted from JavaScript.
 */
internal fun interface ReactNativeBrokerOperationConstructor {
    fun create(appVersion: String, osToken: String): NativeBrokerOperation?
}

internal class AndroidReactNativeBrokerOperationFactory(
    private val constructor: ReactNativeBrokerOperationConstructor =
        LibboxReactNativeBrokerOperationConstructor,
    internal val appVersion: String = BuildConfig.VERSION_NAME,
) : NativeBrokerOperationFactory {
    override fun create(): NativeBrokerOperation {
        val operation = try {
            constructor.create(appVersion, ANDROID_REACT_NATIVE_TOKEN)
        } catch (error: LinkageError) {
            throw unavailableNativeBrokerFailure(error)
        }
        return operation ?: throw unavailableNativeBrokerFailure()
    }

    private companion object {
        const val ANDROID_REACT_NATIVE_TOKEN = "android"
    }
}

/** Production composition point shared only by Android VPN-service clients. */
internal object NativeBrokerTransport {
    val operationFactory: NativeBrokerOperationFactory = AndroidNativeBrokerOperationFactory()
}

/** Production composition point for the dedicated React Native bridge. */
internal object ReactNativeBrokerTransport {
    val operationFactory: NativeBrokerOperationFactory = AndroidReactNativeBrokerOperationFactory()
}

private object LibboxAndroidBrokerOperationConstructor : AndroidBrokerOperationConstructor {
    override fun create(appVersion: String, apiLevel: String): NativeBrokerOperation? {
        val operation = Libbox.newOpenRungBrokerOperationForAndroid(appVersion, apiLevel)
            ?: return null
        return LibboxNativeBrokerOperation(operation)
    }
}

private object LibboxReactNativeBrokerOperationConstructor : ReactNativeBrokerOperationConstructor {
    override fun create(appVersion: String, osToken: String): NativeBrokerOperation? {
        val operation = Libbox.newOpenRungBrokerOperationForReactNative(appVersion, osToken)
            ?: return null
        return LibboxNativeBrokerOperation(operation)
    }
}

/**
 * The only Android adapter that knows generated broker types. Every getter is copied immediately
 * into a platform value on the same background invocation, before the runner closes the operation.
 */
private class LibboxNativeBrokerOperation(
    private val operation: OpenRungBrokerOperation,
) : NativeBrokerOperation {
    override fun firstReachable(
        primary: String,
        limit: Int,
        clientId: String,
        sessionId: String,
    ): NativeBrokerRelayResult? = generatedCall {
        operation.firstReachable(primary, limit, clientId, sessionId)?.snapshot()
    }

    override fun sendTelemetryBatchJSON(
        brokerUrl: String,
        batchJson: String,
    ): NativeBrokerCommonResult? = generatedCall {
        operation.sendTelemetryBatchJSON(brokerUrl, batchJson)?.snapshot()
    }

    override fun runSpeedTest(brokerUrl: String): NativeBrokerSpeedTestResult? = generatedCall {
        operation.runSpeedTest(brokerUrl)?.snapshot()
    }

    override fun fetchManifestCandidate(candidateUrl: String): NativeBrokerManifestResult? =
        generatedCall {
            operation.fetchManifestCandidate(candidateUrl)?.snapshot()
        }

    override fun requestWSSTicket(
        brokerUrl: String,
        relayId: String,
        frontId: String,
        clientId: String,
        sessionId: String,
    ): NativeBrokerWssTicketResult? = generatedCall {
        operation.requestWSSTicket(brokerUrl, relayId, frontId, clientId, sessionId)?.snapshot()
    }

    override fun close() {
        generatedCall { operation.close() }
    }
}

private fun OpenRungBrokerResult.snapshot(): NativeBrokerCommonResult =
    NativeBrokerCommonResult(
        succeeded = succeeded(),
        errorKind = errorKind(),
        errorText = sanitizeNativeErrorText(errorText()),
        httpStatus = httpStatus(),
        retryAfterMillis = retryAfterMillis(),
    )

private fun OpenRungBrokerRelayResult.snapshot(): NativeBrokerRelayResult =
    NativeBrokerRelayResult(
        succeeded = succeeded(),
        errorKind = errorKind(),
        errorText = sanitizeNativeErrorText(errorText()),
        httpStatus = httpStatus(),
        retryAfterMillis = retryAfterMillis(),
        brokerUrl = brokerURL(),
        relayJson = relayJSON(),
        keyId = keyID(),
        signatureVerified = signatureVerified(),
    )

private fun OpenRungBrokerSpeedTestResult.snapshot(): NativeBrokerSpeedTestResult =
    NativeBrokerSpeedTestResult(
        succeeded = succeeded(),
        errorKind = errorKind(),
        errorText = sanitizeNativeErrorText(errorText()),
        httpStatus = httpStatus(),
        retryAfterMillis = retryAfterMillis(),
        bytes = bytes(),
        ttfbMillis = ttfbMillis(),
        downloadDurationMillis = downloadDurationMillis(),
        totalDurationMillis = totalDurationMillis(),
        mbps = mbps(),
    )

private fun OpenRungBrokerManifestResult.snapshot(): NativeBrokerManifestResult =
    NativeBrokerManifestResult(
        succeeded = succeeded(),
        errorKind = errorKind(),
        errorText = sanitizeNativeErrorText(errorText()),
        httpStatus = httpStatus(),
        retryAfterMillis = retryAfterMillis(),
        bodyJson = bodyJSON(),
        sourceUrl = sourceURL(),
    )

private fun OpenRungBrokerWSSTicketResult.snapshot(): NativeBrokerWssTicketResult =
    NativeBrokerWssTicketResult(
        succeeded = succeeded(),
        errorKind = errorKind(),
        errorText = sanitizeNativeErrorText(errorText()),
        httpStatus = httpStatus(),
        retryAfterMillis = retryAfterMillis(),
        ticket = ticket(),
        url = url(),
        expiresAtMillis = expiresAtMillis(),
    )

private inline fun <T> generatedCall(block: () -> T): T = try {
    block()
} catch (error: LinkageError) {
    throw unavailableNativeBrokerFailure(error)
}

/**
 * Runs one synchronous gomobile selector without ever blocking Main.
 *
 * Cancellation dispatches Close to another IO worker before structured concurrency waits for the
 * blocking child. A second idempotent Close always runs from NonCancellable IO. The post-await
 * cancellation check prevents a success racing with cancellation from being committed by callers.
 */
internal suspend fun <T> runNativeBrokerOperation(
    factory: NativeBrokerOperationFactory,
    ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
    call: (NativeBrokerOperation) -> T,
): T {
    val operation = factory.create()
    try {
        return coroutineScope {
            val worker = async(ioDispatcher) { call(operation) }
            try {
                worker.await().also { currentCoroutineContext().ensureActive() }
            } catch (cancellation: CancellationException) {
                withContext(NonCancellable + ioDispatcher) {
                    closeNativeOperation(operation)
                }
                throw cancellation
            }
        }
    } finally {
        withContext(NonCancellable + ioDispatcher) {
            closeNativeOperation(operation)
        }
    }
}

/** Turns an unsuccessful or nil snapshot into the bounded native failure contract. */
internal fun <T : NativeBrokerResult> requireNativeBrokerSuccess(
    result: T?,
    operationName: String,
): T {
    if (result == null) {
        throw BrokerNativeFailure(
            kind = BrokerNativeFailureKind.UNAVAILABLE,
            message = "$operationName returned no native result",
        )
    }
    if (result.succeeded) return result

    val kind = BrokerNativeFailureKind.fromBinding(result.errorKind)
    val detail = sanitizeNativeErrorText(result.errorText)
    val message = if (detail.isEmpty()) {
        "$operationName failed"
    } else {
        "$operationName failed: $detail"
    }
    if (kind == BrokerNativeFailureKind.CANCELLED) {
        throw CancellationException(message)
    }
    throw BrokerNativeFailure(
        kind = kind,
        httpStatus = result.httpStatus.takeIf { it != 0 },
        retryAfterMillis = result.retryAfterMillis.takeIf { it > 0 },
        message = message,
    )
}

internal fun unavailableNativeBrokerFailure(cause: Throwable? = null): BrokerNativeFailure =
    BrokerNativeFailure(
        kind = BrokerNativeFailureKind.UNAVAILABLE,
        message = "native broker transport is unavailable",
        cause = cause,
    )

private fun closeNativeOperation(operation: NativeBrokerOperation) {
    // Preserve the request outcome if a stale artifact also fails during idempotent cleanup.
    runCatching { operation.close() }
}

private const val MAX_NATIVE_ERROR_BYTES = 256

internal fun sanitizeNativeErrorText(value: String?): String {
    val safe = buildString {
        value.orEmpty().forEach { character ->
            when {
                character == '\r' || character == '\n' || character == '\t' -> append(' ')
                character.isISOControl() -> Unit
                else -> append(character)
            }
        }
    }.trim()
    return truncateUtf8(safe, MAX_NATIVE_ERROR_BYTES)
}

private fun sanitizeBrokerFailureMessage(value: String): String =
    truncateUtf8(sanitizeNativeErrorText(value), MAX_NATIVE_ERROR_BYTES)

private fun truncateUtf8(value: String, maxBytes: Int): String {
    val bytes = value.toByteArray(Charsets.UTF_8)
    if (bytes.size <= maxBytes) return value
    var end = maxBytes
    while (end > 0 && (bytes[end].toInt() and 0xC0) == 0x80) end--
    return String(bytes, 0, end, Charsets.UTF_8)
}
